rs/canister_sandbox/src/replica_controller/sandbox_process_eviction.rs

```diff
@@ -1,43 +1,69 @@
 use num_traits::ops::saturating::SaturatingAdd;
 use std::time::Instant;
 
-use ic_types::{CanisterId, NumBytes};
+use ic_types::{AccumulatedPriority, CanisterId, NumBytes};
 
 #[derive(Clone, Eq, PartialEq, Debug)]
 pub(crate) struct EvictionCandidate {
     pub id: CanisterId,
     pub last_used: Instant,
     pub rss: NumBytes,
+    pub scheduler_priority: AccumulatedPriority,
 }
 
-/// Evicts the least recently used candidates in order to bring the number of
-/// the remaining candidates down to `max_count_threshold` and their total RSS
-/// down to `max_sandboxes_rss`.
+/// Evicts the candidates with lowest scheduler priority in order to bring the
+/// number of the remaining candidates down to `max_count_threshold` and their
+/// total RSS down to `max_sandboxes_rss`.
 ///
 /// The function also tries to evict candidates that have been idle for a long
 /// time (`last_used_threshold`).
 pub(crate) fn evict(
-    mut candidates: Vec<EvictionCandidate>,
+    candidates: Vec<EvictionCandidate>,
     total_rss: NumBytes,
     max_count_threshold: usize,
     last_used_threshold: Instant,
     max_sandboxes_rss: NumBytes,
 ) -> Vec<EvictionCandidate> {
-    candidates.sort_by_key(|x| x.last_used);
+    let evict_at_least: usize = candidates.len().saturating_sub(max_count_threshold);
 
-    let evict_at_least = candidates.len().saturating_sub(max_count_threshold);
+    let (mut idle, mut non_idle): (_, Vec<_>) = candidates
+        .into_iter()
+        .partition(|candidate| candidate.last_used < last_used_threshold);
 
-    let mut evicted = vec![];
+    // Evict as many idle candidates as required.
+    idle.sort_by_key(|x| x.last_used);
     let mut evicted_rss = NumBytes::new(0);
+    let mut evicted_num = 0;
 
-    for candidate in candidates.into_iter() {
-        if candidate.last_used >= last_used_threshold
-            && evicted.len() >= evict_at_least
+    for candidate in idle.iter() {
+        if evicted_num >= evict_at_least
             && total_rss <= max_sandboxes_rss.saturating_add(&evicted_rss)
         {
-            // We have already evicted the minimum required number of candidates
-            // and all the remaining candidates were not idle the recent
-            // `last_used_threshold` time window. No need to evict more.
+            break;
+        }
+        evicted_num += 1;
+        evicted_rss = evicted_rss.saturating_add(&candidate.rss);
+    }
+
+    if evicted_num >= evict_at_least && total_rss <= max_sandboxes_rss.saturating_add(&evicted_rss)
+    {
+        // We have already evicted the minimum required number of candidates.
+        // No need to evict more.
+        idle.truncate(evicted_num);
+        return idle;
```

Comment:
You can replace the `break` above with this.

Reply:
That will not give the same behavior because in the case the last idle candidate is added, we will evict at least one non-idle candidate even if all conditions are met.